### PR TITLE
Add support for kubernetes endport field

### PR DESCRIPTION
**Pull request by rikatz**
This PR adds support for the recently added endPort field in kubernetes, which now is feature gated but will be marked as default in next release.

TODO:
* [x] Improve tests
* [x] Validate functioning with Kubernetes
* [x] Verify need to change docs

### CODE REVIEW

**Comment by aauren** [comment]
@rikatz Thanks for thinking of us and submitting this! Definitely appreciate it!

Give me a shout when you finish with the other checkboxes in there and want a review and I'll take another look.

I was looking at the documentation for this feature (https://kubernetes.io/docs/concepts/services-networking/network-policies/#targeting-a-range-of-ports) and I can't tell if this is only for egress, or if `endPort` can show up in both ingress and egress. Did only the `egress` specification change or both?

**Comment by rikatz** [comment]
Hey @aauren for sure. I'm getting some time during this weekend to finish here.

EndPort works for both Ingress and Egress: It's part of the NetworkPolicyPort struct (used in both). In case of Ingress it's more feasible when working with Intra Pod communications (or headless) as services nowadays does not support portRanges (yet! there's a KEP ongoing about that).

https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/2079-network-policy-port-range/README.md Here's the specification if you wanna take a look :)

Thank you very much!

**Comment by rikatz** [comment]
@aauren I've added some unit tests, and tested the correct rule creation in Kubernetes as well. Everything seems to work fine, although I'm running inside KinD and it seems to be complaining about some things.

I'll try to run on a pure kubeadm cluster, but if you can take a look into the PR I would appreciate.

Thanks!!

**Review by aauren — CHANGES_REQUESTED** [review]
Overall it looks really good, thanks for implementing this for us!

I just and a question and two small changes requested.

**Comment by aauren** [comment]
BTW I was also able to test this on my test cluster and validated that the rules were created as we expected.

**Comment by rikatz** [comment]
Hey @aauren sorry for my long delay. I'll cycle back to this probably tomorrow :)

File: pkg/controllers/netpol/policy.go [hunk 1]
```diff
@@ -632,10 +637,25 @@ func (npc *NetworkPolicyController) evalPodPeer(policy *networking.NetworkPolicy
...
 		} else if npPort.Port.Type == intstr.Int {
-			numericPorts = append(numericPorts, protocolAndPort{port: npPort.Port.String(), protocol: string(*npPort.Protocol)})
+			var portproto protocolAndPort
+			if npPort.EndPort != nil {
+				if *npPort.EndPort >= npPort.Port.IntVal {
+					portproto.endport = strconv.Itoa(int(*npPort.EndPort))
+				} else {
+					// Per k8s api validation we should never reach this
+					// but if EndPort is < than Port we should skip
+					continue
```
[the remark this answers is not in the document]
Comment (rikatz) [reply pkg/controllers/netpol/policy.go:654]: Hum, agreed with you! Will change here. Actually the spec drops the netpol object in validation, when this situation happens so in this case this is a last resort here (we assume here that something wrong happened in k8s api validation).

File: pkg/controllers/netpol/network_policy_controller_test.go [hunk 2]
```diff
@@ -395,6 +399,156 @@ func TestNewNetworkPolicySelectors(t *testing.T) {
...
+							},
+						},
+					},
+				},
+			},
+			expectedRule: "-A KUBE-NWPLCY-SQYQ7PVNG6A6Q3DU -m comment --comment \"rule to ACCEPT traffic from source pods to all destinations selected by policy name: simple-egress-pr namespace nsA\" --dport 30000:31000 -j MARK --set-xmark 0x10000/0x10000 \n" +
+				"-A KUBE-NWPLCY-SQYQ7PVNG6A6Q3DU -m comment --comment \"rule to ACCEPT traffic from source pods to all destinations selected by policy name: simple-egress-pr namespace nsA\" --dport 30000:31000 -m mark --mark 0x10000/0x10000 -j RETURN \n" +
+				"-A KUBE-NWPLCY-SQYQ7PVNG6A6Q3DU -m comment --comment \"rule to ACCEPT traffic from source pods to all destinations selected by policy name: simple-egress-pr namespace nsA\" --dport 34000:35000 -j MARK --set-xmark 0x10000/0x10000 \n" +
+				"-A KUBE-NWPLCY-SQYQ7PVNG6A6Q3DU -m comment --comment \"rule to ACCEPT traffic from source pods to all destinations selected by policy name: simple-egress-pr namespace nsA\" --dport 34000:35000 -m mark --mark 0x10000/0x10000 -j RETURN \n",
+		},
```
[the remark this answers is not in the document]
Comment (rikatz) [reply pkg/controllers/netpol/network_policy_controller_test.go:499]: Added

File: pkg/controllers/netpol/policy.go [hunk 3]
```diff
@@ -632,10 +637,25 @@ func (npc *NetworkPolicyController) evalPodPeer(policy *networking.NetworkPolicy
...
+				}
+			}
+			portproto.protocol, portproto.port = protocol, npPort.Port.String()
+			numericPorts = append(numericPorts, portproto)
 		} else {
 			if protocol2eps, ok := namedPort2eps[npPort.Port.String()]; ok {
 				if numericPort2eps, ok := protocol2eps[string(*npPort.Protocol)]; ok {
```
[the remark this answers is not in the document]
Comment (rikatz) [reply pkg/controllers/netpol/policy.go:661]: done :)

**Comment by rikatz** [comment]
@aauren fixed per your comments and already squashed things (as those are minor changes!) 

Sorry once more about the delay, things are pretty busy here :)

**Comment by murali-reddy** [comment]
thanks @rikatz for adding support for port ranges.

@aauren thanks for revewing. I had a chance to look at the PR. Changes LGTM.

**Pull request merged**